src/Pos/Generator/BlockEvent.hs

```diff
@@ -155,360 +344,50 @@ newtype Chance = Chance {getChance :: Rational}
-                    -- The amount of blocks to skip. One is enough.
-                    skip = 1
-                blockIndices <- genBlockIndices (tip + 1 + skip)
-                let
-                    ev = BlkEvApply $ BlockEventApply
-                        { _beaInput    = blockIndices
-                        , _beaOutValid = BlockApplyFailure
-                        }
-                    gbes = GbesExpectedFailure
-                return (Just ev, gbes)
-    -- Success with rollback.
-    genBlockInBlockchain blockchain (IsFailure False) (IsRollback True) = do
-        len <- getRandomR (1, length blockchain)
-        let
-            (blockIndices, remainingBlockchain) = splitAtNewestFirst len $
-                toNewestFirst blockchain
-            -- 'NE.fromList' is valid here because 'len >= 1'.
-            blockIndices' = over _NewestFirst NE.fromList blockIndices
-            remainingBlockchain' = nonEmptyOldestFirst $
-                toOldestFirst remainingBlockchain
-            ev = BlkEvRollback $ BlockEventRollback
-                { _berInput    = blockIndices'
-                , _berOutValid = BlockRollbackSuccess
-                }
-            gbes = maybe GbesStartingState GbesBlockchain remainingBlockchain'
-        return (Just ev, gbes)
-    -- Success without rollback (with apply).
-    genBlockInBlockchain blockchain (IsFailure False) (IsRollback False) = do
-        let tip = NE.last (getOldestFirst blockchain)
-        blockIndices <- genBlockIndices (tip + 1)
-        let
-            ev = BlkEvApply $ BlockEventApply
-                { _beaInput = blockIndices
-                , _beaOutValid = BlockApplySuccess
-                }
-            gbes = GbesBlockchain (blockchain Smg.<> blockIndices)
-        return (Just ev, gbes)
-
-splitAtNewestFirst ::
-    forall a.
-       Int
-    -> NewestFirst NE a
-    -> (NewestFirst [] a, NewestFirst [] a)
-splitAtNewestFirst = coerce (NE.splitAt @a)
-
-nonEmptyOldestFirst ::
-    forall a.
-       OldestFirst [] a
-    -> Maybe (OldestFirst NE a)
-nonEmptyOldestFirst = coerce (nonEmpty @a)
-
-{- NOTE: Reordering corner cases
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-
-When reordering blocks to get an invalid sequence, there's a corner case:
-if there's only one block, it's impossible to get an invalid ordering.
-The reason one might have only one block depends on the means of generation,
-but for 'genBlockIndices' there are two possible causes:
-
-* we rolled 'blockIndexEnd' equal to 'blockIndexStart'
-* 'blockIndexMax' is 1
-
-To avoid dealing with this, we append one more block. The downside is that
-appending this block may cause the sequence to be longer than 'blockIndexMax'.
-
--}
+    checkCount = sum (hhStatusEnd :: HhStatusMap)
+    (hhStatusEnd, revBs') = go Map.empty [] bs
+    bs' = reverse revBs'
+    -- NB. 'go' is tail-recursive.
+    go hhStatusSoFar revNewBs = \case
+        [] -> (hhStatusSoFar, revNewBs)
+        (ev:evs) -> case blkEvTip ev of
```

Comment:
I don't like that you repeat code, i.e. you write `Map.lookup tipHh hhStatusSoFar`.

Indentation is ok, it shows structure. The reason you don't like it is that 4 spaces is too much, with 2 spaces it would look just fine, but codestyle is codestyle.